[하티] 뷰 챌린지 미션 1단계 제출합니다.

sujin9 opened the pull request:
안녕하세요 핑구 🐧 🌼 💕
이번 뷰 챌린지 미션동안 만나서 반가워요!!
잘 부탁 드립니당 ~~~

<image src="https://github.com/woowacourse/android-paint/assets/8490058/79484a92-6edc-4cf8-9255-8498939d77e5" width=360>

이번 1단계 미션을 진행하며 요구 사항들을 열심히 반영했는데요!

다만, 하단에 `색상 변경` `굵기 변경` 의 경우, 
색상 혹은 굵기를 바꿀 때마다 버튼을 누르는 것이 너무 불필요한 작업이라 느껴져서 우선 삭제했습니다.
사실 색상 변경을 누르면 색상 리스트가 보이고, 굵기 변경을 누르면 굵기조절 슬라이드가 보이면서
그림을 그리면 사라지도록 하고 싶었는데,, <- 이 부분을 어떻게 구현해야 할지 감이 잘 안잡혀서 ㅠㅠ 우선 포기했습니다..

연휴 기간 잘 쉬시고 리뷰는 천천히 해주셔도 됩니다!
질문 & DM & 피드백은 언제든 환영이에요 🤗 
그럼 즐거운 추석 연휴 보내요 핑구 ~

sujin9 committed:
feat: main 화면 뷰 (xml) 구현

sujin9 committed:
feat: 컬러 리스트 recyclerview의 item view (xml) 구현

sujin9 committed:
docs: 기능 요구 사항 정리

sujin9 committed:
feat: 자유 곡선 그리기 구현

sujin9 committed:
feat: 브러시 두께 변경 구현

sujin9 committed:
feat: 브러시 색상 변경 구현

sujin9 committed:
feat: 곡선의 양끝이 둥글고 부드럽도록 구현

sujin9 committed:
feat: 하단 버튼 제거

sujin9 committed:
refactor: data binding 적용

sujin9 committed:
refactor: 함수 분리 및 리팩토링

sujin9 committed:
docs: 기능 요구 사항 만족한 것 표시

sujin9 committed:
refactor: 색상 recyclerview 성능 개선

pingu244 commented on app/src/main/java/woowacourse/paint/main/ColorViewHolder.kt:
```diff
@@ -0,0 +1,33 @@
+class ColorViewHolder(
+    parent: ViewGroup,
+    onColorClickListener: (Int) -> Unit,
+) : RecyclerView.ViewHolder(
+    LayoutInflater.from(parent.context).inflate(
+        R.layout.item_color,
+        parent,
+        false,
+    ),
```
뷰 홀더를 만들 때 이렇게 만드는 것은 처음 보는데 어떤 방식으로 작동하는지 설명해주실 수 있을까요?

sujin9 replied:
adapter에서 createViewHolder를 할 때 부모 view를 넘기구, viewHolder에서 itemView를 넘기는 곳에서 바로 inflate를 해주어 itemView 객체를 만들어요!

리사이클러뷰에 들어가는 각각의 아이템 뷰를 관리하는 곳인 ViewHolder 내에서 inflate를 해주는 것이 맞다고 생각해서 이렇게 구현했어요 ;)

pingu244 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,9 +1,45 @@
 <?xml version="1.0" encoding="utf-8"?>
-<androidx.constraintlayout.widget.ConstraintLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
```
데이터 바인딩을 사용하는 이유가 있을까요? 여기서는 뷰바인딩만 사용해도 괜찮다고 보입니다

sujin9 replied:
data binding을 사용할 가능성을 열어두고 싶기도 했고, 습관적으로 적용해준 이유도 없지않아 있는 것 같네요,, 😅 
현재는 코드 수정을 통해 data binding을 활용하고 있답니다! :)

pingu244 commented on app/src/main/java/woowacourse/paint/main/MainActivity.kt:
```diff
@@ -0,0 +1,46 @@
+        binding.rangeSliderBrushSize.values = listOf(DEFAULT_BRUSH_SIZE)
+        binding.ctvPaintBoard.apply {
+            setBrushSize(DEFAULT_BRUSH_SIZE)
+            setBrushColor(getColor(PaintColor.values().first().colorRes))
```
`getColor(PaintColor.values().first().colorRes)` 이 부분이 너무 긴 것 같아요..! PaintColor 클래스에서 정의해놓은 getColorRes() 함수를 사용하시는 것은 어떨까요?

sujin9 replied:
좋은 의견 같아요!! 👍 
미리 함수를 만들어두고 안쓰고 있었네요 ㅋㅋ
그리고 현재는 아예 Default color값(첫 번째 색)을 가져올 수 있도록 수정했습니다! :)

pingu244 commented on app/src/main/java/woowacourse/paint/main/PaintBoard.kt:
```diff
@@ -0,0 +1,75 @@
+    private fun setDefaultPaint() {
+        currentPaint.apply {
+            strokeCap = Paint.Cap.ROUND
+            strokeJoin = Paint.Join.ROUND
```
제 그림판에서는 왜이리 삐쭉빼쭉하지 했는데 여기서 힌트를 얻네요!! 감사합니다 하티😉

pingu244 commented on app/src/main/java/woowacourse/paint/main/PaintBoard.kt:
```diff
@@ -0,0 +1,75 @@
+        currentPaint.apply {
+            strokeCap = Paint.Cap.ROUND
+            strokeJoin = Paint.Join.ROUND
+            isAntiAlias = true
```
이 친구는 무슨 역할을 해주는 건가요?

sujin9 replied:
설명드릴게요 핑구! 😆 🌼 
(열심히 찾아보았답니다 ㅋㅋ)
혹시 틀린 부분이 있거나, 이해가 되지 않는 부분이 있다면 말씀해주세요 ~ 💘 

---

## strokeCap
선의 양 끝 선 처리를 해줍니다.

`strokeCap = Paint.Cap.ROUND` 를 주석처리한 경우에는, 다음과 같이 보여요.
<img src="https://github.com/woowacourse/android-paint/assets/8490058/05d691c8-25b2-458e-9cd5-9cf45479983d" width=200>

ref.
https://developer.android.com/reference/android/graphics/Paint#setStrokeCap(android.graphics.Paint.Cap)
https://developer.android.com/reference/android/graphics/Paint.Cap

## strokeJoin
선의 꼭지점 처리를 해줍니다.

`strokeJoin = Paint.Join.ROUND` 를 주석처리한 경우에는, 다음과 같이 보여요.
<img src="https://github.com/woowacourse/android-paint/assets/8490058/d26364c2-46ab-4931-816d-94a8366f8b33" width=200>


ref.
https://developer.android.com/reference/android/graphics/Paint#setStrokeJoin(android.graphics.Paint.Join)
https://developer.android.com/reference/android/graphics/Paint.Join

## isAntiAlias (setAntiAlias)
true로 설정하는 경우, 선 자체에는 영향이 없고, 선의 가장자리를 부드럽게 해줍니다.
주변 배경과 잘 어울리도록 하는 효과가 있다고 해요 ;)

![image](https://github.com/woowacourse/android-paint/assets/8490058/7c104234-5f44-4ca4-b5b7-2b699be9dcae)

ref.
https://developer.android.com/reference/android/graphics/Paint#isAntiAlias()

pingu244 replied:
정성스러운 답변 정말 감사해요..❣️

pingu244 commented on app/src/main/java/woowacourse/paint/main/PaintBoard.kt:
```diff
@@ -0,0 +1,75 @@
+    @SuppressLint("ClickableViewAccessibility")
+    override fun onTouchEvent(event: MotionEvent): Boolean {
+        when (event.action) {
+            MotionEvent.ACTION_DOWN -> startDrawing(event)
+            MotionEvent.ACTION_MOVE -> moveDrawing(event)
+            MotionEvent.ACTION_UP -> endDrawing()
```
함수화를 아주 잘 해주셨군요 굿굿

pingu244 commented on app/src/main/java/woowacourse/paint/main/PaintBoard.kt:
```diff
@@ -0,0 +1,75 @@
+    private fun startDrawing(event: MotionEvent) {
+        currentPath = Path()
+        currentPath.moveTo(event.x, event.y)
+        currentPath.lineTo(event.x, event.y)
```
moveDrawing()에서 lineTo로 그려주니 시작할 때는 그냥 moveTo만 해줘도 되지 않을까요?

sujin9 replied:
우선 저는 화면 위로 드래그를 하면 선이 그려지고, 그냥 터치만 하는 경우에는 점이 찍히도록 구현하고자 했어요.
시작할 때도 `lineTo`를 해주지 않으면, 화면을 그냥 터치했을 때 점이 찍히지 않더라구요 :(
그래서 해당 라인을 추가하게 되었습니다..!

혹시 핑구도 이런 문제를 겪으셨나요?
해결하셨다면, 어떤 방법으로 하셨는지 궁금해요! :)

pingu244 replied:
오 그런 부분에 대해서는 생각을 못했네요..! 하티 덕분에 알아가는게 많아지네요 감사합니다 ><

pingu244 commented on app/src/main/java/woowacourse/paint/model/PaintColor.kt:
```diff
@@ -0,0 +1,19 @@
+import woowacourse.paint.R
+
+enum class PaintColor(@ColorRes val colorRes: Int) {
+    RED(R.color.red),
+    ORANGE(R.color.orange),
+    YELLOW(R.color.yellow),
+    GREEN(R.color.green),
+    BLUE(R.color.blue),
```
저는 그냥 ColorRes값이 없는 enum class로 정의해주고 Custom View에서 색상을 지정해주었는데 이렇게 한 곳에서 관리해주는 것이 더 좋은 것 같아요!👍

pingu244 reviewed (changes requested):
하티 안녕하세요~!
리뷰이로 만나게 되어 반가워요🥰✨🐤🐥
너무 늦게 리뷰드리게 되어 죄송합니다..
전체적으로 코드들이 깔끔해서 리뷰하기 좋았어요! 궁금한 점 위주로 코멘트 남겨놓았으니 한 번 확인부탁드려요~😁
<image src="https://github.com/woowacourse/android-paint/assets/69796976/570b9d2b-dc88-420a-989c-0ec7295e36fb" width=360>

sujin9 committed:
feat: viewModel 및 dataBinding 적용

sujin9 committed:
refactor: RangeSlider를 Slider로 변경

- RangeSlider의 경우, 양쪽 값을 설정해주어야 하는데 굳이 해당 기능이 필요 없다고 판단
- value를 통해 초기값을 설정해줄 수 있음

sujin9 committed:
feat: 색상들 중 첫번째가 항상 default 색이 되도록 설정

sujin9 committed:
refactor: 패키지 수정

sujin9 committed:
refactor: nullable이 필요없는 곳 non-nullable로 수정

sujin9 committed:
feat: 화면 세로모드로 고정

pingu244 commented on app/src/main/java/woowacourse/paint/main/PaintBoard.kt:
```diff
@@ -8,9 +8,11 @@ import android.graphics.Path
 import woowacourse.paint.model.DrawablePathHistory
+import woowacourse.paint.model.PaintColor
 
 class PaintBoard constructor(context: Context, attrs: AttributeSet) : View(context, attrs) {
```
현재는 xml 상에서만 PaintBoard를 사용할 수 있어요! 동적 생성을 하려면 어떻게 해야할까요?

pingu244 reviewed (approved):
리뷰 답변과 반영 고생하셨습니다~!~!
하티의 코드를 리뷰하면서 오히려 제가 많이 얻어가네요😊
마지막까지 꼼꼼하게 코드들 정비해주셔서 더이상 드릴 리뷰가 없는 것 같아요!
코멘트 하나 남겼는데 생각해보시고 다음 미션에서 반영해주세요~
바로 머지하겠습니다 ;) 다음 단계도 화이팅👍
